Fleshing out Execution and Memory models

llvm-beanz opened the pull request:
This update Intro.Defs earlier in the introduction and puts more detail into the SPMD programming model and memory model.

The PDF draft of this change is available [here](https://github.com/microsoft/hlsl-specs/files/12552783/hlsl.pdf).

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -54,6 +54,47 @@
+to return unsuccessfully and produce one or more diagnostic messages.
+
+\Sub{Implementation-defined Behavior}{Intro.Defs.ImpDef}
+\p Behavior of a well formed program and correct data which may vary by the
```
Nit: I think "well-formed" is a more accepted spelling, similar to your use of "ill-formed" elsewhere.

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -54,6 +54,47 @@
+implementation, and the implementation is expected to document the behavior.
+
+\Sub{Implementation Limits}{Intro.Defs.ImpLimits}
+\p Restrictions imposed upon programs by the implementation.
```
I wonder if it's worth clarifying what "the implemention" refers to, since we have both DXC/Clang, but also downstream consumers of the DXIL output.

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -54,6 +54,47 @@
+\p Restrictions imposed upon programs by the implementation.
+
+\Sub{Undefined Behavior}{Intro.Defs.Undefined}
+\p Behavior of invalid program constructs or incorrect data which this standard
```
wording: "... or incorrect data _for which_ this standard ..."

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -54,6 +54,47 @@
+imposes no requirements, or does not sufficiently detail.
+
+\Sub{Unspecified Behavior}{Intro.Defs.Unspecified}
+\p Behavior of a well formed program and correct data which may vary by the
```
What does "correct data" refer to here? I would consider rephrasing to something like "Undocumented implementation-dependent behavior of a well-formed program" (a bit more compact)

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -54,6 +54,47 @@
+
+\Sub{Well-formed Program}{Intro.Defs.WellFormed} \p An \acrshort{hlsl} program
+constructed according to the syntax rules, diagnosable semantic rules, and the
+One Definition Rule.
```
Interesting that ODR is called out as a separate rule, are there no other rules in the same classification (link-time) that need to be considered for a program to be considered "well-formed"?

llvm-beanz replied:
It's possible in the future we may need to incorporate more post-compile rules. I actually expect that we may accrue a few to represent the DXIL and SPIR-V validators, but I want to hold off referencing that explicitly until I get a little further along.

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -77,46 +118,67 @@
 execution, however it differs in one key way. In multi-threaded programming
 threads advance independent of each other. In \acrshort{spmd} programs, a group
 of \gls{lane}s execute instructions in lock step because each instruction is a
 \acrshort{simd} instruction computing the results for multiple \gls{lane}s
```
wording suggestion:

> In \acrshort{spmd} programs, a group of \gls{lane}s share a common instruction stream, which they execute in lockstep.

Note: I believe "lockstep" is commonly used as a single word now, although it is occasionally hyphenated also. This wording also includes instructions that don't "compute results" (e.g. barrier instructions).

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -77,46 +118,67 @@
+\Par{\gls{wave}}{Intro.Model.Terms.Wave} A grouping of \gls{lane}s for execution
+is called a \gls{wave}. \gls{wave} sizes vary by hardware architecture. Some
+hardware implementations support multiple wave sizes. Generally wave sizes are
+powers of two, but there is no requirement that be the case. \acrshort{hlsl} is
```
Is there a "multiple-of-four" requirement?

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -77,46 +118,67 @@
 
-\Sub{\gls{threadgroup}}{Intro.Model.Group}
+\Par{\gls{threadgroup}}{Intro.Model.Terms.Group}
+A grouping of \gls{lane}s executing the same shader to produce a combined
```
It may be worth explicitly clarifying the execution model expected when a thread group size is less than the wave size, and/or not a multiple of the wave size.

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -77,46 +118,67 @@
+A grouping of \gls{lane}s executing the same shader to produce a combined
+result is called a \gls{threadgroup}. \gls{threadgroup}s are independent of
+\acrshort{simd} hardware specifications. The dimensions of a \gls{threadgroup}
+are defined in three dimensions. The maximum dimensions of a \gls{threadgroup}
```
wording: I suggest "the maximum extent along each dimension of a ..." because "maximum dimensions" here is just 3.

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -77,46 +118,67 @@
+\Sub{\acrshort{spmd} Execution Model}{Intro.Model.Exec}
+
+\p A runtime implementation will provide an implementation-defined mechanism for
+defining a \gls{dispatch}. In an implementation-defined way a runtime will
```
I might rephrase this a bit, although I'm not sure what the best phrasing is. I think I would naturally write something like:

> Implementations are permitted to vary with respect to hardware resource management and execution scheduling provided that the observed behavior conforms to the behavior described in this specification. For example, a runtime implementation may choose to...

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -130,37 +192,38 @@
+Device memory is memory available to all \gls{lane}s executing on the device.
+This memory may be read or written to by multiple \gls{threadgroup}s that are
+executing concurrently. Reads and writes to device memory may occur in any order
+except as restricted by synchronization intrinsics. Some device memory may be
```
"synchronization intrinsics" might preclude specific annotations like `globallycoherent` which I don't personally consider to be an "intrinsic"

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -130,37 +192,38 @@
+executing concurrently. Reads and writes to device memory may occur in any order
+except as restricted by synchronization intrinsics. Some device memory may be
+visible to the host. Device memory that is visible to the host may have
+additional synchronization concerns.
```
Probably worth elaborating what those concerns are specifically

jeremyong commented on specs/language/introduction.tex:
```diff
@@ -130,37 +192,38 @@
+during execution.
+
+\Par{Host Memory}{Intro.Memory.Spaces.Host}
+Host memory is memory that is directly addressable by the host. Reads and writes
```
What about resizable bar and host-addressable VRAM?

jeremyong reviewed (commented):
Thanks for working on this doc! I left some wording suggestions and comments that are hopefully helpful.

llvm-beanz commented:
I've put in a bunch of word smithing and updates.

I think this is mostly in sane state now.

[Updated PDF](https://github.com/microsoft/hlsl-specs/files/13529743/hlsl.pdf)

llvm-beanz commented on specs/language/introduction.tex:
```diff
@@ -130,37 +241,38 @@
-\gls{isoC} and \gls{isoCPP} specifications, however they are included here for
-reader convenience.
+\p \acrshort{hlsl} programs manipulate data stored in five distinct memory
+spaces: thread, threadgroup, device, constant and host.
```
I had stripped `host` out and missed that spot. We define a host address space in the DXIL spec, but we don't use it, so I was thinking I'd simplify this until we really need it.

llvm-beanz commented:
@Keenuts, the last change I added here is in response to the discussion we had in email this week. Do you think that wording correctly addresses the concerns?

Keenuts commented on specs/language/introduction.tex:
```diff
@@ -78,46 +128,144 @@
+
+\p A \textit{inactive} \gls{lane} is a lane that is not executed by the initial
+launch conditions. This occurs when there is not sufficient inputs to fill the
+all \gls{lane}s in the \gls{wave}.
```
Maybe not worth to mention, but isn't another cause is register pressure? If the register pressure is too high, AFAIK lanes might be disabled and the wave size be only have a fraction of the theoretical wave-size.

Keenuts commented:
> @Keenuts, the last change I added here is in response to the discussion we had in email this week. Do you think that wording correctly addresses the concerns?

Sure, this block as-is makes sense: no optimization should change the program behavior (except optimizations like fast-math? But maybe that's considered to be a codegen switch, not an optimization switch?)

What is IMO the important missing bit (but it's a whole subject) is to define what is expected for actual execution: the spec mentions lock-step is not mandatory, hence the spec as-is is allows the same CFG/shader to yield 2 different results. Example: running on Maxwell and Volta architecture, with the latter allowing not reconverging directly after a divergence.

llvm-beanz commented:
> Sure, this block as-is makes sense: no optimization should change the program behavior (except optimizations like fast-math? But maybe that's considered to be a codegen switch, not an optimization switch?)

Within the spec you can't change the program behavior outside what is specified. HLSL has fast-math by default, so that will be the specified behavior for floating point math operations.

> What is IMO the important missing bit (but it's a whole subject) is to define what is expected for actual execution: the spec mentions lock-step is not mandatory, hence the spec as-is is allows the same CFG/shader to yield 2 different results. Example: running on Maxwell and Volta architecture, with the latter allowing not reconverging directly after a divergence.

Lockstep isn't strictly required by the spec for anything. Even wave operations can be emulated without lockstep execution if you have appropriate synchronization primitives.

For Volta, the expectation is that wave operations will effectively require warp synchronization. I think that's a driver problem for Nvidia more than anything else. Maybe that's the missing explicit wording.

I can add something like:

> Wave, Quad, and Thread Group operations require execution synchronization of applicable active and helper Lanes as defined by the individual operation.

Keenuts commented:
> For Volta, the expectation is that wave operations will effectively require warp synchronization. I think that's a driver problem for Nvidia more than anything else. Maybe that's the missing explicit wording.
> 
> I can add something like:
> 
> > Wave, Quad, and Thread Group operations require execution synchronization of applicable active and helper Lanes as defined by the individual operation.

Yes, and where those barriers are put should probably be described in the control-flow structures (here using Cuda's syntax mixed with HLSL)

```
switch (value) {
    case 0:
        WaveActiveSum(1);
   default:
       __syncwrap(MASK_FOR_ALL_THREADS);
       WaveActiveSum(10);
}
```

llvm-beanz commented:
> ```
> switch (value) {
>     case 0:
>         WaveActiveSum(1);
>    default:
>        __syncwrap(MASK_FOR_ALL_THREADS);
>        WaveActiveSum(10);
> }
> ```

I don't think this is correct. Every `Wave*` intrinsic needs to have a `__syncwarp` associated with it. In your example it may have only been required for the `default:` case, but in a more complicated example they're effectively needed for all Wave and Quad operations with a mask that represents all the participating lanes.

Keenuts commented:
> > ```
> > switch (value) {
> >     case 0:
> >         WaveActiveSum(1);
> >    default:
> >        __syncwrap(MASK_FOR_ALL_THREADS);
> >        WaveActiveSum(10);
> > }
> > ```
> 
> I don't think this is correct. Every `Wave*` intrinsic needs to have a `__syncwarp` associated with it. In your example it may have only been required for the `default:` case, but in a more complicated example they're effectively needed for all Wave and Quad operations with a mask that represents all the participating lanes.

Yes, of course. If we add a `case 1` which doesn't branch, then the mask has to be different, what I meant with this is the control-flow structure itself needs to define where synchronization is expected, not just the intrinsic:
 - we both agree the we must synchronize as soon as we can, when we fall-through. But this has to be made explicit IMO, and it's related to the control flow construct.
 (Same for loops breaks, etc).
 
 But It seems this goes beyond this initial PR, so probably not the best place to dive into this.

llvm-beanz commented:
Yea, I think some of this will come down to the wording as we write the behavior of switch statements and case labels, but at the same time I don't think we need to define synchronization for control flow structures.

It isn't illegal to reshape control flow during optimization if it doesn't change program behavior, so synchronization of execution isn't really a property of the control flow structures. Even within the definitions of C, it is illegal to merge and hoist control flow if it changes the resulting execution. C doesn't have the SIMT/SPMD considerations that HLSL does, but we can lean on similar rules for HLSL and get the expected behavior.

The C spec defined case statements from ANSI C through until C17 (the wording is slightly different in the latest draft of C23) with the following statement:

> Labels in themselves do not alter the flow of control, which continues unimpeded across them.

If that's the basis for how `case` labels are defined, and `Wave*` operations are defined to have wave sync behavior, I think that defines the expected behavior completely.

I don't want our spec to make statements about expectations for control flow structure that aren't required because we do want conforming compilers to be allowed to optimize in cases where they can.

damyanp commented on specs/language/introduction.tex:
```diff
@@ -78,46 +128,148 @@
+\SubSub{\gls{wave}}{Intro.Model.Terms.Wave}
 
-\Sub{\gls{wave}}{Intro.Model.Wave}
+\p A grouping of \gls{lane}s for execution is called a \gls{wave}. The size of a
```
All lanes in a wave will execute in lockstep (right?) Wonder if that can be mentioned here.  Text above suggests that lanes are only running in lockstep when there's some kind of synchronization going on.

llvm-beanz replied:
They aren't required to be lock-step except at synchronization points. While it has been true with a lot of hardware that lock-step is guaranteed, that's less true with newer hardware. NV's latest GPUs have per-thread program counters which can allow much more creative divergence in execution.

llvm-beanz commented:
I've marked the outstanding conversations as resolved. I think that I've addressed all the feedback. I'm happy to revisit anything in subsequent PRs. I'm sure this will all need some more work as we get further along.
